File: java/eyes.playwright.java/src/main/java/com/applitools/eyes/playwright/universal/Refer.java
package com.applitools.eyes.playwright.universal;

import com.applitools.eyes.universal.Reference;

import java.util.HashSet;
import java.util.UUID;

public class Refer extends com.applitools.eyes.universal.Refer {

    public Refer() {
        super();
    }

    /**
     * store the driver's ref.
     *
     * @return the UUID
     */
    public String ref(Object value, Reference root) {
        if (value == null) {
            return null;
        }

        String ref = UUID.randomUUID().toString();
        references.put(ref, value);

        if (root != null) {
            HashSet<Reference> childRefs = this.relations.computeIfAbsent(root.getApplitoolsRefId(), k -> new HashSet<>());
            childRefs.add(new Reference(ref));
        }
        return ref;
    }

    /**
     * get a ref from store
     *
     * @param ref  the ref
     * @return the ref
     */
    public Object deref(Object ref) {
        if (isRef(ref)) {
            Reference reference = (Reference) ref;
            return references.get(reference.getApplitoolsRefId());
        }

        return ref;
    }

    public void destroy(Reference root) {
        if (!isRef(root)) {
            return;
        }

        HashSet<Reference> childRefs = relations.get(root.getApplitoolsRefId());
        if (childRefs != null) {
            for (Reference childRef : childRefs) {
                destroy(childRef);
            }

            relations.remove(root.getApplitoolsRefId());
        }

        references.remove(root.getApplitoolsRefId());
    }

    public Boolean isRef(Object ref) {
        if (ref instanceof Reference)
            return ((Reference) ref).getApplitoolsRefId() != null;

        return false;
    }

}

File: dotnet/Eyes.Playwright.DotNet/Universal/Refer.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using Applitools.Universal;

namespace Applitools.Playwright.Universal
{
    public class Refer : Applitools.Universal.Refer
    {
        private readonly ConcurrentDictionary<string, object> references_ = new ConcurrentDictionary<string, object>();

        private readonly ConcurrentDictionary<string, HashSet<Reference>> relations_ =
            new ConcurrentDictionary<string, HashSet<Reference>>();

        ///<summary>store the driver's ref.</summary>
        ///<returns>the ref GUID</returns>
        public string Ref(object value, Reference root)
        {
            if (value == null)
            {
                return null;
            }

            string refId = Guid.NewGuid().ToString();
            references_.TryAdd(refId, value);

            if (root != null)
            {
                if (!relations_.TryGetValue(root.ApplitoolsRefId, out HashSet<Reference> childRefs))
                {
                    relations_[root.ApplitoolsRefId] = childRefs = new HashSet<Reference>();
                }

                childRefs.Add(new Reference(refId));
            }

            return refId;
        }

        ///<summary>get a ref from store</summary>
        ///<param name="ref">the ref</param>
        /// <returns>the ref</returns>
        public object Deref(object @ref)
        {
            if (IsRef(@ref))
            {
                Reference reference = (Reference)@ref;
                references_.TryGetValue(reference.ApplitoolsRefId, out object result);
                return result;
            }

            return @ref;
        }

        public override void Destroy(Reference root)
        {
            if (!IsRef(root))
            {
                return;
            }

            relations_.TryGetValue(root.ApplitoolsRefId, out HashSet<Reference> childRefs);
            if (childRefs != null)
            {
                foreach (Reference childRef in childRefs)
                {
                    Destroy(childRef);
                }

                relations_.TryRemove(root.ApplitoolsRefId, out _);
            }

            references_.TryRemove(root.ApplitoolsRefId, out _);
        }

        private static bool IsRef(object @ref)
        {
            return @ref is Reference reference && reference.ApplitoolsRefId != null;
        }
    }
}